pkg/virt-launcher/virtwrap/cmd-server/server.go

```diff
@@ -0,0 +1,237 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package cmdserver
+
+import (
+	"encoding/json"
+	"net"
+	"net/rpc"
+	"os"
+	"path/filepath"
+
+	k8sv1 "k8s.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/log"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/cmd-server/client"
+)
+
+type Launcher struct {
+	domainManager virtwrap.DomainManager
+}
+
+func getK8SecretsfromClientArgs(args *cmdclient.Args) (map[string]*k8sv1.Secret, error) {
+	var secrets map[string]*k8sv1.Secret
+	err := json.Unmarshal([]byte(args.K8SecretMapJSON), &secrets)
+	if err != nil {
+		log.Log.Reason(err).Errorf("Failed to unmarshal k8 secrents json object")
+		return nil, err
+	}
+	return secrets, nil
+}
+
+func getVmfromClientArgs(args *cmdclient.Args) (*v1.VirtualMachine, error) {
+	vm := &v1.VirtualMachine{}
+	err := json.Unmarshal([]byte(args.VMJSON), vm)
+	if err != nil {
+		log.Log.Reason(err).Errorf("Failed to unmarshal vm json object")
+		return nil, err
+	}
+	return vm, nil
+}
+
+func (s *Launcher) SyncSecret(args *cmdclient.Args, reply *cmdclient.Reply) error {
+	reply.Success = true
+
+	vm, err := getVmfromClientArgs(args)
+	if err != nil {
+		reply.Success = false
+		reply.Message = err.Error()
+		return nil
+	}
+
+	err = s.domainManager.SyncVMSecret(vm,
+		args.SecretUsageType,
+		args.SecretUsageID,
+		args.SecretValue)
+
+	if err != nil {
+		log.Log.Object(vm).Reason(err).Errorf("Failed to sync vm secrets")
+		reply.Success = false
+		reply.Message = err.Error()
+		return nil
+	}
+
+	log.Log.Object(vm).Info("Synced vm secrets")
+	return nil
+}
+
+func (s *Launcher) Start(args *cmdclient.Args, reply *cmdclient.Reply) error {
```

Comment:
I think that `Start` is misleading. Having something like `Sync` in the name is more correct. Also right now, virt-handler would try to unpause a VM, if it is in an unexpected pause mode, so to some degree it really already tries to synchronize the vm and the domain state.

Reply:
fixed